master/internal/resourcemanagers/priority.go

```diff
@@ -0,0 +1,331 @@
+package resourcemanagers
+
+import (
+	"fmt"
+	"sort"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/determined-ai/determined/master/pkg/actor"
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+type priorityScheduler struct {
+	preemptionEnabled bool
+}
+
+// NewPriorityScheduler creates a new scheduler that schedules tasks via priority.
+func NewPriorityScheduler(config *SchedulerConfig) Scheduler {
+	return &priorityScheduler{preemptionEnabled: config.Priority.Preemption}
+}
+
+func (p *priorityScheduler) Schedule(rp *ResourcePool) ([]*AllocateRequest, []*actor.Ref) {
+	return p.prioritySchedule(rp.taskList, rp.groups, rp.agents, rp.fittingMethod)
+}
+
+func (p *priorityScheduler) prioritySchedule(
+	taskList *taskList,
+	groups map[*actor.Ref]*group,
+	agents map[*actor.Ref]*agentState,
+	fittingMethod SoftConstraint,
+) ([]*AllocateRequest, []*actor.Ref) {
+	agentsSplitByLabel := splitAgentsByLabel(agents)
+	toAllocate := make([]*AllocateRequest, 0)
+	toRelease := make([]*actor.Ref, 0)
+
+	// Since labels are a hard scheduling constraint, process every
+	// label independently.
+	for label, agentsWithLabel := range agentsSplitByLabel {
+		toAllocatedForLabel, toReleaseForLabel := p.priorityScheduleByLabel(
+			taskList, groups, agentsWithLabel, fittingMethod, label)
+		toAllocate = append(toAllocate, toAllocatedForLabel...)
+		toRelease = append(toRelease, toReleaseForLabel...)
+	}
+
+	return toAllocate, toRelease
+}
+
+func (p *priorityScheduler) priorityScheduleByLabel(
+	taskList *taskList,
+	groups map[*actor.Ref]*group,
+	agents map[*actor.Ref]*agentState,
+	fittingMethod SoftConstraint,
+	label string,
+) ([]*AllocateRequest, []*actor.Ref) {
+	// All pending zero slot tasks get scheduled right away.
+	toAllocate := getAllPendingZeroSlotTasks(taskList, label)
+	toRelease := make(map[*actor.Ref]bool)
+
+	// Sort tasks by priorities and timestamps. This sort determines the order in which
+	// tasks are scheduled and preempted.
+	priorityToPendingTasksMap, priorityToScheduledTaskMap := sortTasksByPriorityAndTimestamp(
+		taskList, groups, label)
+
+	// Make a local copy of the agent state that we will modify.
+	localAgentsState := deepCopyAgents(agents)
+
+	// Once we are unable to start a task of a higher priority, do not start anymore tasks.
```

Comment:
q: does this mean we should wait for all higher priority tasks to be scheduled before scheduling lower-priority tasks even if there are available slots for lower-priority tasks? It looks the last behavior is for the situation I am describing.

Reply:
This was explicitly discussed in the ERD but I agree it's a good idea to add it to the PR description as well, going to do that.

Reply:
Thanks for adding that to the memory. We definitely can find the initial plan in the ERD or PRD but I found it very hard when I tried to refresh the memory on some specific lines code if I had to go to the documentation.